### PR TITLE
Regional weather diffusion dataloader updates

### PR DESCRIPTION
**Pull request by pzharrington**
# PhysicsNeMo Pull Request

## Description
Modifies the recipe dataset contract to also support `physicsnemo.datapipes` style data loaders, along with some functionality allowing multi-region training (e.g. CONUS+EU). Also modifies the policy to only load the target under `"state"` key, and require users to always put the condition in the `"background"` key (including preceding state timesteps).

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/weather/regional_weather_diffusion/utils/parallel.py [hunk 1]
```diff
@@ -102,72 +166,141 @@ def local_batch_size(self, global_batch_size: int) -> int:
...
-                yield from local_samples
-
-        return torch.utils.data.DataLoader(
-            dataset=dataset,
+        local_seed = None if seed is None else seed + self.dist.rank
+        rng = np.random.default_rng(seed=local_seed)
+        while True:
+            if shuffle:
+                rng.shuffle(local_samples)
+            yield from local_samples
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply examples/weather/regional_weather_diffusion/utils/parallel.py:245]: Fixed in [be11cb9](https://github.com/NVIDIA/physicsnemo/pull/1996/commits/be11cb97d1f793c9a4c9bd8975469fa0774b5fe5)

File: examples/weather/regional_weather_diffusion/datasets/__init__.py [hunk 2]
```diff
@@ -13,23 +13,31 @@
...
+    flatten_group_collate,
+)
+from .mock import MockDataset
 
+# StormCastDataSource implementations keyed by "<module>.<ClassName>", matching
+# the `dataset.name` config field.
+dataset_classes: dict[str, type[StormCastDataSource]] = {
+    "data_loader_hrrr_era5.HrrrEra5Dataset": HrrrEra5Dataset,
+    "mock.MockDataset": MockDataset,
+}
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply examples/weather/regional_weather_diffusion/datasets/__init__.py]: Fixed in `be11cb9`

**Comment by pzharrington** [comment]
@greptileai
